components/batch-system/src/batch.rs

```diff
@@ -303,30 +284,53 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
                     batch.release_control(&self.router.control_box, len);
                 }
             }
-            if !batch.normals.is_empty() {
-                for (i, p) in batch.normals.iter_mut().enumerate() {
-                    let len = self.handler.handle_normal(p);
-                    batch.counters[i] += 1;
-                    if p.is_stopped() {
-                        reschedule_fsms.push((i, ReschedulePolicy::Remove));
-                    } else {
-                        if batch.counters[i] > 3 {
-                            hot_fsm_count += 1;
-                            // We should only reschedule a half of the hot regions, otherwise,
-                            // it's possible all the hot regions are fetched in a batch the
-                            // next time.
-                            if hot_fsm_count % 2 == 0 {
-                                reschedule_fsms.push((i, ReschedulePolicy::Schedule));
-                                continue;
-                            }
-                        }
-                        if let Some(l) = len {
-                            reschedule_fsms.push((i, ReschedulePolicy::Release(l)));
+
+            let mut hot_fsm_count = 0;
+            for (i, p) in batch.normals.iter_mut().enumerate() {
+                let len = self.handler.handle_normal(p);
+                batch.counters[i] += 1;
+                if p.is_stopped() {
+                    reschedule_fsms.push((i, ReschedulePolicy::Remove));
+                } else {
+                    if batch.counters[i] > 3 {
+                        hot_fsm_count += 1;
+                        // We should only reschedule a half of the hot regions, otherwise,
+                        // it's possible all the hot regions are fetched in a batch the
+                        // next time.
+                        if hot_fsm_count % 2 == 0 {
+                            reschedule_fsms.push((i, ReschedulePolicy::Schedule));
+                            continue;
                         }
                     }
+                    if let Some(l) = len {
+                        reschedule_fsms.push((i, ReschedulePolicy::Release(l)));
+                    }
+                }
+            }
+            let mut fsm_cnt = batch.normals.len();
+            while batch.normals.len() < max_batch_size {
```

Comment:
Shall we remove `reschedule_fsms` from `batch.normals` before this step?

Reply:
No. Because there will also be other Fsms need to be reschedule. And remove fsms should keep order. But it is a good idea that we shall release fsm as soon as possible. Maybe in next PR we can improve this point.